File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int mod = 1000000007;
    long max = 0;
    long total;
    public int getTotal(TreeNode node) {
        if(node==null) return 0;
        //MODIFYING TREES NODES VAL TO TOTAL VAL OF SUBTREE FOR TRAVERSE FUNCTION
        return node.val+= getTotal(node.left) + getTotal(node.right);
    }
    
    public void traverse(TreeNode node) {
        if(node==null) return;
        traverse(node.left);
        traverse(node.right);
        max = Math.max(node.val*(total-node.val),max);
    }
    public int maxProduct(TreeNode root) {
        this.total = getTotal(root);
        traverse(root);
        return (int)(max%mod);
    }
}

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int mod = 1000000007;
    long maxi = 0;
    long total=0;
    int convertTree(TreeNode* node) {
        if(!node) return 0;
        return node->val += convertTree(node->left) + convertTree(node->right);
    }  
    void traverse(TreeNode* node) {
        if(!node) return;
        traverse(node->right);
        traverse(node->left);
        maxi = max(node->val*(total-node->val),maxi);
    }
    int maxProduct(TreeNode* root) {
        this->total = convertTree(root);   
        traverse(root);
        return (int)(maxi%mod);
    }
};